### PR TITLE
Schema list tables

### PR DESCRIPTION
**Pull request by Nic0tiN**
Fixes issues #10847 and #10854.

### CODE REVIEW

File: src/Database/Schema/MysqlSchema.php [hunk 1]
```diff
@@ -25,9 +25,22 @@ class MysqlSchema extends BaseSchema
...
      * {@inheritDoc}
      */
-    public function listTablesSql($config)
+    public function listTablesSql($config, $type = 'both')
     {
-        return ['SHOW TABLES FROM ' . $this->_driver->quoteIdentifier($config['database']), []];
+        $table_type = null;
+        $sql = 'SHOW FULL TABLES FROM ' . $this->_driver->quoteIdentifier($config['database']);
+
+        if ($type != 'both') {
```
Comment (markstory) [inline src/Database/Schema/MysqlSchema.php]: Can we have constants for these string values? Constants allow for typo free method calls.

File: src/Database/Schema/SqliteSchema.php [hunk 2]
```diff
@@ -105,13 +105,26 @@ protected function _convertColumn($column)
...
+            $sql .= ' AND type = ?';
+
+            if ($type == 'tables') {
+                $params = ['table'];
+            } else {
+                $params = ['view'];
+            }
+        } else {
+            $sql .= ' AND (type = ? OR type = ?)';
+            $params = ['table', 'view'];
```
Comment (markstory) [inline src/Database/Schema/SqliteSchema.php]: Isn't this `AND TYPE IN (?, ?)`?

File: src/Database/Schema/SqlserverSchema.php [hunk 3]
```diff
@@ -25,16 +25,32 @@ class SqlserverSchema extends BaseSchema
...
+        if ($type != 'both') {
+            $sql .= ' AND TABLE_TYPE = ?';
+
+            if ($type == 'tables') {
+                $params = ['BASE TABLE'];
+            } else {
+                $params = ['VIEW'];
+            }
+        } else {
+            $sql .= ' AND (TABLE_TYPE = ? OR TABLE_TYPE = ?)';
```
Comment (markstory) [inline src/Database/Schema/SqlserverSchema.php]: Another place you could use `IN`.

**Review by markstory — COMMENTED** [review]
There are a few lint errors that need to be fixed up.

**Comment by ADmad** [comment]
Shouldn't this target 3.next due to signature change to listTablesSql()?

File: src/Database/Schema/MysqlSchema.php (hunk 1 above)
Comment (dereuromark) [inline src/Database/Schema/MysqlSchema.php]: Was about to comment the same: class constants would be best here.

File: src/Database/Schema/TableSchema.php [hunk 4]
```diff
@@ -278,6 +278,27 @@ class TableSchema
     const ACTION_SET_DEFAULT = 'setDefault';
 
     /**
+     * Table type both
+     *
+     * @var string
+     */
+    const TABLE_TYPE_BOTH = 'both';
```
Comment (dereuromark) [inline src/Database/Schema/TableSchema.php]: should this be maybe REPOSITORY_TYPE_* ?

File: phpunit.xml.dist [hunk 5]
```diff
@@ -51,7 +51,7 @@
         <env name="db_dsn" value="mysql://localhost/cake_test?timezone=UTC"/>
         -->
         <!-- SQL Server
-        <env name="db_dsn" value="sqlserver://localhost/cake_test?timezone=UTC"/>
+        <env name="db_dsn" value="sqlserver://localhost\localhost/cake_test?timezone=UTC"/>
```
Comment (ADmad) [inline phpunit.xml.dist]: This needs to be undone.

File: src/Database/Schema/Collection.php [hunk 6]
```diff
@@ -53,13 +53,14 @@ public function __construct(ConnectionInterface $connection)
     }
 
     /**
-     * Get the list of tables available in the current connection.
-     *
+     * Get the list of tables or views available in the current connection.
+     * @param string $tableType Type of objects. Possible values are:
+     * TableSchema::TABLE_TYPE_VIEW, TableSchema::TABLE_TYPE_TABLE or TableSchema::TABLE_TYPE_BOTH (default).
```
Comment (markstory) [inline src/Database/Schema/Collection.php]: Do these constants exist? The code uses `REPOSITORY_TYPE_*`

File: src/Database/Schema/SqlserverSchema.php [hunk 7]
```diff
@@ -25,16 +25,32 @@ class SqlserverSchema extends BaseSchema
     /**
      * {@inheritDoc}
      */
-    public function listTablesSql($config)
+    public function listTablesSql($config, $type = 'both')
```
Comment (markstory) [inline src/Database/Schema/SqlserverSchema.php]: Missing a constant here.

File: src/Database/Schema/MysqlSchema.php [hunk 8]
```diff
@@ -25,9 +25,22 @@ class MysqlSchema extends BaseSchema
...
      * {@inheritDoc}
      */
-    public function listTablesSql($config)
+    public function listTablesSql($config, $type = TableSchema::REPOSITORY_TYPE_BOTH)
     {
-        return ['SHOW TABLES FROM ' . $this->_driver->quoteIdentifier($config['database']), []];
+        $tableType = null;
+        $sql = 'SHOW FULL TABLES FROM ' . $this->_driver->quoteIdentifier($config['database']);
+
+        if ($type != TableSchema::REPOSITORY_TYPE_BOTH) {
```
Comment (dereuromark) [inline src/Database/Schema/MysqlSchema.php]: `!=`

File: src/Database/Schema/MysqlSchema.php [hunk 9]
```diff
@@ -25,9 +25,22 @@ class MysqlSchema extends BaseSchema
...
+            $sql .= ' WHERE Table_type = ?';
+
+            if ($type == TableSchema::REPOSITORY_TYPE_TABLE) {
```
Comment (dereuromark) [inline src/Database/Schema/MysqlSchema.php]: `===`

**Comment by dereuromark** [comment]
With a few small fixes this could go soon into 3.6 then.

**Comment by ADmad** [comment]
Closing for reasons stated by markstory.

**Pull request closed without merging**